File: pkcs11wrapper-tests/app/src/androidTest/java/ru/rutoken/pkcs11wrapper/attribute/IPkcs11AttributeFactoryTest.java
package ru.rutoken.pkcs11wrapper.attribute;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import ru.rutoken.pkcs11wrapper.attribute.longvalue.Pkcs11LongAttribute;
import ru.rutoken.pkcs11wrapper.attribute.longvalue.Pkcs11ObjectClassAttribute;
import ru.rutoken.pkcs11wrapper.constant.IPkcs11AttributeType;
import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11AttributeType;
import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11ObjectClass;
import ru.rutoken.pkcs11wrapper.rule.ModuleRule;

/**
 * We have to use instrumented tests because of unit tests problems with native libraries.
 */
@RunWith(AndroidJUnit4.class)
public class IPkcs11AttributeFactoryTest {
    private static final ModuleRule sModule = new ModuleRule();

    private static IPkcs11AttributeFactory getFactory() {
        return sModule.getValue().getAttributeFactory();
    }

    @Test
    public void makeAttribute() {
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_CLASS);
        assertTrue("wrong type", attribute instanceof Pkcs11ObjectClassAttribute);

        var longAttribute = getFactory()
                .makeAttribute(Pkcs11LongAttribute.class, Pkcs11AttributeType.CKA_CLASS);
        assertTrue("wrong type", longAttribute instanceof Pkcs11ObjectClassAttribute);

        var classAttribute = getFactory()
                .makeAttribute(Pkcs11ObjectClassAttribute.class, Pkcs11AttributeType.CKA_CLASS);
    }

    @Test
    public void makeAttributeSetBooleanValue() {
        final var value = false;
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_TOKEN, value);
        assertTrue("wrong type", attribute instanceof Pkcs11BooleanAttribute);
        assertEquals("value not equal", attribute.getValue(), value);
    }

    @Test
    public void makeAttributeSetStringValue() {
        var value = "test";
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_LABEL, value);
        assertTrue("wrong type", attribute instanceof Pkcs11StringAttribute);
        assertEquals("value not equal", attribute.getValue(), value);
    }

    @Test
    public void makeAttributeSetIntValue() {
        //int is not long
        var value = 0;
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_CERTIFICATE_CATEGORY, value);
        assertTrue("wrong type", attribute instanceof Pkcs11LongAttribute);
        assertEquals("value not equal", attribute.getValue(), (long) value);
    }

    @Test
    public void makeAttributeSetLongValue() {
        var value = 0L;
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_CERTIFICATE_CATEGORY, value);
        assertTrue("wrong type", attribute instanceof Pkcs11LongAttribute);
        assertEquals("value not equal", attribute.getValue(), value);
    }

    @Test
    public void makeAttributeSetEnumValue() {
        var value = Pkcs11ObjectClass.CKO_PRIVATE_KEY;
        var attribute = getFactory().makeAttribute(Pkcs11AttributeType.CKA_CLASS, value);
        assertTrue("wrong type", attribute instanceof Pkcs11ObjectClassAttribute);
        assertEquals("value not equal", attribute.getValue(), value.getAsLong());
    }

    @Test(expected = ClassCastException.class)
    public void makeAttributeWrongClass_negative() {
        var classAttribute = getFactory()
                .makeAttribute(Pkcs11ObjectClassAttribute.class, Pkcs11AttributeType.CKA_KEY_TYPE);
    }

    @Test
    public void makeAllAttributes() {
        for (IPkcs11AttributeType type : Pkcs11AttributeType.values()) {
            getFactory().makeAttribute(type);
        }
    }
}

File: pkcs11wrapper-tests/app/src/androidTest/java/ru/rutoken/pkcs11wrapper/attribute/IPkcs11AttributeFactoryTest.kt
/*
 * Copyright (c) 2018, Aktiv-Soft JSC.
 * See the LICENSE file at the top-level directory of this distribution.
 * All Rights Reserved.
 */

package ru.rutoken.pkcs11wrapper.attribute

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.equality.shouldBeEqualToComparingFields
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import org.junit.ClassRule
import org.junit.Test
import ru.rutoken.pkcs11wrapper.attribute.longvalue.*
import ru.rutoken.pkcs11wrapper.constant.AttributeLongValueSupplier
import ru.rutoken.pkcs11wrapper.constant.IPkcs11AttributeType
import ru.rutoken.pkcs11wrapper.constant.standard.*
import ru.rutoken.pkcs11wrapper.datatype.Pkcs11Date
import ru.rutoken.pkcs11wrapper.rule.highlevel.ModuleRule

class IPkcs11AttributeFactoryTest {
    @Test
    fun makeAttribute() {
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_CLASS)
        attribute.shouldBeInstanceOf<Pkcs11ObjectClassAttribute>()

        val longAttribute = factory.makeAttribute(Pkcs11LongAttribute::class.java, Pkcs11AttributeType.CKA_CLASS)
        longAttribute.shouldBeInstanceOf<Pkcs11ObjectClassAttribute>()

        val classAttribute =
            factory.makeAttribute(Pkcs11ObjectClassAttribute::class.java, Pkcs11AttributeType.CKA_CLASS)
        classAttribute.shouldBeInstanceOf<Pkcs11ObjectClassAttribute>()
    }

    @Test
    fun makeAttributeSetBooleanValue() {
        val value = false
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_TOKEN, value)
        attribute.shouldBeInstanceOf<Pkcs11BooleanAttribute>()
        attribute.value shouldBe value
    }

    @Test
    fun makeAttributeSetStringValue() {
        val value = "test"
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_LABEL, value)
        attribute.shouldBeInstanceOf<Pkcs11StringAttribute>()
        attribute.value shouldBe value
    }

    @Test
    fun makeAttributeSetIntValue() {
        val value = 1
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_CERTIFICATE_CATEGORY, value)
        attribute.shouldBeInstanceOf<Pkcs11LongAttribute>()
        attribute.value shouldBe value.toLong()
    }

    @Test
    fun makeAttributeSetLongValue() {
        val value = 1L
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_CERTIFICATE_CATEGORY, value)
        attribute.shouldBeInstanceOf<Pkcs11LongAttribute>()
        attribute.value shouldBe value
    }

    @Test
    fun makeAttributeSetByteArrayValue() {
        val value = byteArrayOf(0x01, 0x02, 0x03)
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_VALUE, value)
        attribute.shouldBeInstanceOf<Pkcs11ByteArrayAttribute>()
        attribute.value shouldBe value
    }

    @Test
    fun makeAttributeSetDateValue() {
        val value = Pkcs11Date(
            module.value.lowLevelFactory.makeDate().apply {
                year = "2023".toByteArray()
                month = "01".toByteArray()
                day = "17".toByteArray()
            }
        )
        val attribute = factory.makeAttribute(Pkcs11AttributeType.CKA_START_DATE, value)
        attribute.shouldBeInstanceOf<Pkcs11DateAttribute>()
        attribute.value shouldBeEqualToComparingFields value
    }

    @Test
    fun makeAttributeSetEnumValue() {
        testMakeAttributeSetEnumValue<Pkcs11ObjectClassAttribute>(
            Pkcs11AttributeType.CKA_CLASS, Pkcs11ObjectClass.CKO_PRIVATE_KEY
        )

        testMakeAttributeSetEnumValue<Pkcs11CertificateTypeAttribute>(
            Pkcs11AttributeType.CKA_CERTIFICATE_TYPE, Pkcs11CertificateType.CKC_X_509
        )

        testMakeAttributeSetEnumValue<Pkcs11MechanismTypeAttribute>(
            Pkcs11AttributeType.CKA_NAME_HASH_ALGORITHM, Pkcs11MechanismType.CKM_SHA256,
        )

        testMakeAttributeSetEnumValue<Pkcs11KeyTypeAttribute>(
            Pkcs11AttributeType.CKA_KEY_TYPE, Pkcs11KeyType.CKK_GOSTR3410
        )
    }

    @Test
    fun makeAttributeWrongClass_negative() {
        shouldThrow<ClassCastException> {
            factory.makeAttribute(Pkcs11ObjectClassAttribute::class.java, Pkcs11AttributeType.CKA_KEY_TYPE)
        }
    }

    @Test
    fun makeAllAttributes() {
        Pkcs11AttributeType.values().forEach { type -> factory.makeAttribute(type) }
    }

    private inline fun <reified T : Any> testMakeAttributeSetEnumValue(
        type: IPkcs11AttributeType,
        value: AttributeLongValueSupplier
    ) {
        factory.makeAttribute(type, value).let { attribute ->
            attribute.shouldBeInstanceOf<T>()
            attribute.value shouldBe value.asLong
        }
    }

    companion object {
        @ClassRule
        @JvmField
        val module = ModuleRule()

        private val factory = module.value.attributeFactory
    }
}